util/feeding.py

```diff
@@ -111,20 +138,36 @@ def sparse_reshape(sparse):
         shape = sparse.dense_shape
         return tf.sparse.reshape(sparse, [shape[0], shape[2]])
 
-    def batch_fn(wav_filenames, features, features_len, transcripts):
+    def batch_fn(wav_filenames, features, features_len, transcripts, review_audios):
         features = tf.data.Dataset.zip((features, features_len))
         features = features.padded_batch(batch_size,
                                          padded_shapes=([None, Config.n_input], []))
         transcripts = transcripts.batch(batch_size).map(sparse_reshape)
         wav_filenames = wav_filenames.batch(batch_size)
-        return tf.data.Dataset.zip((wav_filenames, features, transcripts))
+
+        # In order not to waste too much prefetch performance, randomly extract only `one` audio for each step
+        if FLAGS.augmentation_review_audio_steps and batch_size > 1:
+            skip_size = tf.random.uniform(shape=[], minval=0, maxval=batch_size - 1, dtype=tf.int64)
+            review_audio = review_audios.skip(skip_size).batch(1)
+        else:
+            review_audio = review_audios.batch(1)
+
+        return tf.data.Dataset.zip((wav_filenames, features, transcripts, review_audio))
 
     num_gpus = len(Config.available_devices)
-    process_fn = partial(entry_to_features, train_phase=train_phase)
 
-    dataset = (tf.data.Dataset.from_generator(generate_values,
-                                              output_types=(tf.string, (tf.int64, tf.int32, tf.int64)))
-                              .map(process_fn, num_parallel_calls=tf.data.experimental.AUTOTUNE))
+    if noise_dirs_or_files:
+        noise_iterator = create_noise_iterator(noise_dirs_or_files, read_csvs)
+    else:
+        noise_iterator = None
+
+    process_fn = partial(entry_to_features, train_phase=train_phase, noise_iterator=noise_iterator)
```

Comment:
The noise dataset has been `.shuffle() + .repeat()`, if the `dataset = dataset.cache(cache_path)` doesn't execute, it should mix the different noise on same speech file with every loop.
I have double-checked it, using tf.Print to inspect the speech/noise filenames for each epoch, and make sure that the noises were selected randomly.
If we review the audio on tensorboard, we will notice that the same noise file is always picked, which is due to the random seed, try changing the `random_seed` to get a different result.

Reply:
Yes, but how about caching training samples and noise separately?

Reply:
I didn't plan to cache noise, because mixing audio happens in `entry_to_features()` then converted to features, caching audio could cost 10 times memory than caching features, which should be less practical and should not speed up the training too much, so I made `enable_cache=False` when `FLAG.train_augmentation_files` is not null like other augmentation does.